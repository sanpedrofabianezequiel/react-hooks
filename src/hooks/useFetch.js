import { useEffect } from "react"
import { useState } from "react/cjs/react.development"


export const useFetch = (url)=>{
    const [state, setState] = useState({data:null,loading:true,error:null})

    useEffect(() => {

        //limpiamos para ver el Loading
        setState({data:null,loading:true,error:null});

        fetch(url)
            .then( x => x.json())
            .then( data =>{
                setState({
                    loading:false,
                    error:null,
                    data:data
                });
            })
            .catch(()=>{
                setState({
                    data:null,
                    loading:false,
                    error:"no se pudo cargar la info"
                });
            })

    }, [url]);
    return state;
}