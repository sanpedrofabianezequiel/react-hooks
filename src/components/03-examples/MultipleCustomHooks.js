import React from 'react'
import { useFetch } from '../../hooks/useFetch';
import '../02-useEffect/effects.css';
import {useCounter} from '../../hooks/useCounter';


export const MultipleCustomHooks = () => {

    const {state,increment}  = useCounter(1); //state is COUNTER
    const {loading,data} =useFetch(`https://breakingbadapi.com/api/quotes/${state}`);
    const {author,quote} = !!data && data[0];

    return (
        <div>
            <h1>Breaking Bad Quotes</h1>
            <hr className="mb-10"/>

            {
                (loading) 
                ? (
                    <div className="alert alert-info text-center">
                      Loading...
                    </div>
                  )
                :(
                    <blockquote className="blockquote text-right">
                      <p className="mb-10">{quote}</p>
                      <footer className="blockquote-footer">{author}</footer>
                    </blockquote>
                 )
            }
            
            <button className="btn btn-primary"
                    onClick={increment}
            >
                Next Card
            </button>

        </div>
    )
}
