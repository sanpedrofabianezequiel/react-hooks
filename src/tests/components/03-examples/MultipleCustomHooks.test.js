import React from 'react';
import {shallow} from 'enzyme';
import '@testing-library/jest-dom';
import { MultipleCustomHooks } from '../../../components/03-examples/MultipleCustomHooks';
import {useFetch } from '../../../hooks/useFetch';
import { useCounter } from '../../../hooks/useCounter';
jest.mock('../../../hooks/useFetch');



describe('Pruebas en <MultipleCustomHooks />', () => {

    
   

    test('debe de mostrarse correctamente', () => {

        //Como aca se llama al componente por primera vez necesitamos inicializar el useFetch
        //con la data " que va a devolver"
        useFetch.mockReturnValue({
            data:null,
            loading:true,
            error:null
        });


        const wrapper= shallow(<MultipleCustomHooks/>);
        expect(wrapper).toMatchSnapshot();
    });
    
    test('shoul debe de mostar la informacion', () => {
        
        //Como aca se llama al componente por primera vez necesitamos inicializar el useFetch
        //con la data " que va a devolver"
        useFetch.mockReturnValue({
            data:[{
                author:"Ezequiel",
                quote:"Hola mundo"
            }],
            loading:false,
            error:null
        });

        const wrapper= shallow(<MultipleCustomHooks/>);
        //console.log(wrapper.html());

        expect(wrapper.find('.alert').exists() ).toBe(false);
        expect(wrapper.find('p').text().trim()).toBe('Hola mundo');
        expect(wrapper.find('footer').text().trim()).toBe('Ezequiel');

    })
    
})
