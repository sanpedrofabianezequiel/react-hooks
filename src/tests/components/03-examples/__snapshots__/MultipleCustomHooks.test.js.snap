// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <MultipleCustomHooks /> debe de mostrarse correctamente 1`] = `
<div>
  <h1>
    Breaking Bad Quotes
  </h1>
  <hr
    className="mb-10"
  />
  <div
    className="alert alert-info text-center"
  >
    Loading...
  </div>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Next Card
  </button>
</div>
`;
